### ISSUE TITLE
Nullable Reference Types annotations

### ISSUE THREAD

**Issue by jnm2** [nunit/nunit#3376]
tl;dr Now that C# 8 is out, people will be using the Nullable Reference Types language feature. We should consider putting the compiler annotations on our library. That would require moving the codebase to C# 8. To reap the benefits in our own codebase, we could also use https://github.com/tunnelvisionlabs/ReferenceAssemblyAnnotator.

### What it is

People who put `<Nullable>enable</Nullable>` in their projects or `#nullable enable` in their source files are then allowed to use the syntax `ReferenceType?` where nulls are permitted. In typical software, nulls end up being actually intended in only a small percentage of the usages of reference types.

The compiler then produces warnings (we have /warnaserror) when a thing that might be null is treated as though it can't be null, like when dotting off which could cause an NRE or when passing to a method which could cause an ArgumentNullException, or when setting a field or property which only causes a problem at a later point in time.

This goes a long way toward solving the billion-dollar problem of nullability blindness in the type system.

https://docs.microsoft.com/en-us/dotnet/csharp/nullable-references
https://devblogs.microsoft.com/dotnet/nullable-reference-types-in-csharp/
https://docs.microsoft.com/en-us/dotnet/csharp/tutorials/nullable-reference-types

Things like `string.IsNullOrEmpty` are annotated with new compiler attributes such as `[NotNullWhen(false)]` on its parameter so that the compiler can recognize that it's a null check. These attributes can be defined by anyone as internal so that they don't conflict with other assemblies. .NET Core 3.0 is the first version of the BCL that adds these attributes as public classes. Referencing https://github.com/tunnelvisionlabs/ReferenceAssemblyAnnotator would define these internally for us without having to maintain the internal attribute definitions ourselves.

### Reasons to do it

- It would be nice to annotate our library so that people who use NUnit are warned if they forget to check things like `TestContext.CurrentContext` for null or if they pass things that could be null to our APIs. It's being up-front about the types that each API accepts in the same way that declaring a parameter `IFoo foo` instead of `object foo` is being more up-front and helpful.

- This would aid the mission of maintaining high-quality internal code written by many different people. We've had our share of null reference exceptions that nothing had forced us to notice at compile time.

- It's not a breaking change.

### Reasons not to do it

- It requires `<LangVersion>8<LangVersion>`. While Visual Studio wouldn't be required in order to compile the codebase, 2019 Update 3 would become the minimum version if you *were* using Visual Studio. You could also compile using the .NET Core 3.0 SDK using the command line or using VS Code and other IDEs.

- It's likely to require us to think about things we never had to reconcile before. Once we start shining around this black light, so to speak, we might not like some of the things we see. (Or is this good, as NUnit 4 begins to take shape?)

- It would touch a decent percentage of our files.

### Proposal

I think we should do it in v3 (because this is my function in life, apparently?) and it's work I enjoy and would prioritize. However, I see these downsides and I'm not going to be overly disappointed if we don't want to do it in v3.

I feel more strongly that we should do this by v4. Otherwise, as one of the most-frequently used libraries, I think we'll end up standing out in an unappealing way as time goes on. I could easily be wrong. But if we do this by v4, the easiest way might be to start now in v3 since it can all be done incrementally.

**Comment by ChrisMaddock** [comment]
I would love to see this in place in v3 also. I personally think the benefits of doing this outweigh the unfortunate cost of limiting the IDEs here.

**Comment by Dreamescaper** [comment]
Could you clarify what is `ReferenceAssemblyAnnotator`? According to description, it is much more than simple defining attributes, right?
Just suggesting an alternative - we could use https://github.com/manuelroemer/Nullable package, it seems to be much more lightweight.

Another thing I'd like to ask. Apart from annotating actual nullable properties, should we think about nullability attributes on assertions somehow? E.g.
```csharp
string? str = GetNullable();
Assert.That(str, Is.Not.Null);
Assert.That(str.Length, Is.Not.Zero);  // <----- should we care that no warning here?
```

**Comment by jnm2** [comment]
@Dreamescaper https://github.com/manuelroemer/Nullable only defines internal attributes in your source. ReferenceAssemblyAnnotator does that but more importantly it rewrites all the .NET Framework or .NET Standard reference assemblies that you compile against to retroactively add nullability attributes by copying them from .NET Core.

> Another thing I'd like to ask. Apart from annotating actual nullable properties, should we think about nullability attributes on assertions somehow? E.g.

The current annotations are too limited for the compiler to be able to tell that `Assert.That(str, Is.Not.Null);` is a null check. We could approach it from the NUnit.Analyzers angle, though.

**Comment by Dreamescaper** [comment]
@jnm2 
> ReferenceAssemblyAnnotator does that too, but more importantly it rewrites all the .NET Framework or .NET Standard reference assemblies that you compile against to retroactively add nullability attributes by copying them from .NET Core.

Yeah. So it makes sense if you cannot update your project for whatever reason. Why is that needed, however, if we can simply add standard2.1 target, where it would simply work?


> The current annotations are too limited for the compiler to be able to tell that Assert.That(str, Is.Not.Null); is a null check. We could approach it from the NUnit.Analyzers angle, though.

We _technically_ could introduce some `That` overloads for `NullConstraint` and `NotNullConstraint`, and add corresponding attributes there.
But I agree - it makes more sense to do from analyzer perspective.
Probably via `DiagnosticSuppressor` ?

**Comment by jnm2** [comment]
> Yeah. So it makes sense if you cannot update your project for whatever reason. Why is that needed, however, if we can simply add netcoreapp3.0 target, where it would simply work?

I don't remember if we are planning to add a netcoreapp3.0 target, but I know we intend to keep netstandard2.0 and net472 targets in addition to netcoreapp2.1. Also, if we're doing this in v3, we're keeping the net35 target.

Without `ReferenceAssemblyAnnotator`, this means that the NUnit codebase would get warnings in the code editor if we select `netcoreapp3.0` in the project context dropdown in the top left of the code editor, but no warnings in the code editor if we select `netcoreapp2.1`, `netstandard2.0`, or `net472`. And that's if we even add a netcoreapp3.0 target.

**Comment by jnm2** [comment]
@Dreamescaper Did I answer what you were asking? I might be misunderstanding, sorry. If we use ReferenceAssemblyAnnotator, it will be impossible to tell downstream. It doesn't go in the nuspec or add an extra assembly. It just makes the compiler smarter when this codebase compiles.

**Comment by jnm2** [comment]
@nunit/framework-team I played around with this since I have my head in the space at work and other places. I'm collecting places where we have bad assumptions by not checking for null when we really should have been the whole time, or by having complex timing dependencies between methods that make it hard to reason about the state that the class fields are in.

Should I start submitting standalone fixes using C# 7 in the meantime?

**Comment by jnm2** [comment]
Okay, thanks for the input.

My instinct is against a long-lived branch because merging is always painful and people will have to be aware of both threads. For now, I'm so close to having NUnit fully annotated that I think I'll tackle opening the issues/PRs. I'll have questions and PRs that I'll need help with.

**Comment by woodstoke** [comment]
We need not forget specify [DoesNotReturn] attributes. 
Now, I'm writing code like this to prevent warnings:

```
if (!OK())
{
    Assert.Fail("my error message");
    throw new Exception();
}
```

**Comment by jnm2** [comment]
We can't use `[DoesNotReturn]` because Assert.Fail does not always throw.

The difference is whether you are inside an Assert.Multiple block.

**Comment by jnm2** [comment]
Also, even when `[DoesNotReturn]` is applied, C# still requires the throw afterwards. The C# language design team decided to make the nullability attributes affect only the nullability state of variables, not flow control. So this would be the only difference:

```cs
nullableThing = a?.B;
if (nullableThing is null) Assert.Fail("Some message");
nullableThing.Foo(); // Warns that `nullableThing` may be null unless Assert.Fail has [DoesNotReturn]
```

**Comment by ddshpak** [comment]
Very glad to see this is being worked on! Thanks!

**Comment by loop-evgeny** [comment]
Is there an estimate of when this is expected to be released? We're enabling nullable reference types in our projects now and the lack of NUnit support for them is making this difficult.

**Comment by jons-aura** [comment]
I'm trying to enable nullable reference types as well and having to drop null forgiving operators "!" even after Is.Not.Null checks is doable but rather irksome so it would be great to see this released.

**Comment by jnm2** [comment]
Some of you are going to be disappointed. Assertions do not always throw when the assertion fails due to the multiple asserts feature, so even `Assert.Null` can't be annotated the way you'd expect. Further, even if that wasn't a concern, the C# language does not provide a mechanism to annotate `Assert.That(obj, Is.Not.Null);` as having anything to do with nullability.

**Comment by jons-aura** [comment]
I am disappointed that it sounds like an explicit `Assert.NotNull()` can't be annotated as a null check as I'm sure a lot of us hoped. Not surprised about `Assert.That` being impossible to annotate though both based on how it works and the comments I saw about it farther up the thread.

**Comment by jnm2** [comment]
One thing https://github.com/nunit/nunit.analyzers could explore is using a diagnostic suppressor automatically for `Assert.That(obj, Is.Not.Null);` and `Assert.NotNull(obj);` when not inside a multiple assert block. It wouldn't be perfect and it could take some time to implement, but we welcome contributions.

(@mikkelbu, we could access the https://docs.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.diagnostics.diagnosticsuppressor.reportsuppressions API using reflection since you want to continue supporting VS 2017.)

**Comment by jnm2** [comment]
One thing https://github.com/nunit/nunit.analyzers could explore is using a diagnostic suppressor automatically for `Assert.That(obj, Is.Not.Null);` and `Assert.NotNull(obj);` when not inside a multiple assert block. It wouldn't be perfect and it could take some time to implement, but we welcome contributions.

(@mikkelbu, we'd have to implement https://docs.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.diagnostics.diagnosticsuppressor in a separate, conditionally-loaded assembly that compiles against Roslyn 3.3.1+ since you want to continue supporting VS 2017.)

**Comment by jnm2** [comment]
Another option is to use https://www.nuget.org/packages/Shouldly. I've been using it in all my NUnit test projects for a couple of years, and I added full nullability annotations in the recently-released v4. Example:

```cs
var foo = someComplexNullableExpression.ShouldNotBeNull();
foo.Bar.ShouldBe(42);
foo.GetItems().ShouldBeEmpty();
```

**Comment by manfred-brands** [comment]
@jons-aura We can annotate `Assert.NotNull` and `Assert.IsNotNull` using the `NotNullAttribute`. This indicates to the compiler that the parameter will not have the value `null` when exiting the function. I just tested this and it works.

``
Assert.NotNull(possiblyNull);
Assert.That(possiblyNull.Length, Is.GreaterThanOrEqualTo(0));
```

The other thing I do a lot is assign a nullable field/return to a non-nullable local with help of an extension method:
```
string? possiblyNull = SomeMethodThatMightReturnNull();
string reallyNotNull = possiblyNull.AssertNotNull();
```

I created PR #3677 for this, so you can try it out.

**Comment by manfred-brands** [comment]
We can add the same `NotNull` attribute to those overloads of `Assert.That` which accept an actual Value.
This assumes that all `IResolveContraint` likely use the value.

```
Assert.That(possiblyNull, Is.Not.Null);
Assert.That(possiblyNull.Length, Is.GreaterThanOrEqualTo(0));
```

This fails for the one constraint that actually tests for `null`, but no code would try to dereference that variable afterwards as the developer assumed it to be `null`.

Updated the PR with an extra commit.

**Comment by jnm2** [comment]
@manfred-brands The team decided at https://github.com/nunit/nunit/issues/3473#issuecomment-592254847 that we cannot add that constraint to Assert.NotNull because Assert.NotNull does not always throw when it fails due to the multiple asserts feature.

I missed linking this in my previous comment, but there's already an open issue for conditionally suppressing nullability warnings after Assert.NotNull in NUnit.Analyzers: https://github.com/nunit/nunit.analyzers/issues/157

**Comment by jnm2** [comment]
A point I'd like to add is that the BCL and C# language designers do not see getting rid of `!` as a goal. It's intended to be used often and long-term, whenever the framework compiler can't prove what you're doing is safe. You need it after passing a path you know is rooted to System.IO.Path.GetDirectoryName. Seeing `!` in the source code gives you useful information. It says "a human vouched for something that the compiler doesn't know."

**Comment by manfred-brands** [comment]
> (@mikkelbu, we'd have to implement https://docs.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.diagnostics.diagnosticsuppressor in a separate, conditionally-loaded assembly that compiles against Roslyn 3.3.1+ since you want to continue supporting VS 2017.)

I will have a play with this.

**Comment by manfred-brands** [comment]
@jnm2 Raised https://github.com/nunit/nunit.analyzers/pull/325

**Comment by rprouse** [comment]
I had a long discussion with @jnm2 about this in Slack. I'm okay with changing the legacy asserts in this way, just not adding new ones, so that part isn't a problem. My issue is that this is a partial fix that cannot have a complete fix within the framework. If we do this for `Assert.NotNull` in the framework, then people will expect it to work for `Assert.That(actual, Is.NotNull)` and might also expect it to work in multiple assert blocks.

I think that we could handle all of these things in the Analyzer project, so it is the best place for that. Unfortunately, there are also some technical difficulties there, so let's leave this PR open for now and see how the work goes in the Analyzer project.

https://github.com/nunit/nunit.analyzers/issues/157

Thanks for working on this in the Analyzers project @manfred-brands 👍🏻

**Comment by mikkelbu** [comment]
Note that @manfred-brands work is part of the 3.0.0 release of the NUnit Analyzers - https://www.nuget.org/packages/NUnit.Analyzers/3.0.0. This implements the initial suppression of errors due to nullable reference types (suppressed by a check via `Assert.NotNull`, `Assert.IsNotNull` or `Assert.That`). The release also contains suppression of the "nullability" (if that is a word) of the exception type returned from `Assert.Throws`.

**Comment by egraff** [comment]
I've just pulled the latest NUnit, and noticed that my ReSharper is giving a lot of warnings when I use test patterns such as

```csharp
InvalidOperationException ex = Assert.Throws<InvalidOperationException>(
    () => DoSomething()
);

Assert.That(ex.Message, Is.EqualTo("Some error message"));
```

where it complains that dereferencing `ex.Message` is a possible `NullReferenceException`.

I've looked at the annotations introduced in https://github.com/nunit/nunit/commit/2e4ebf7d38e1e8e16689f06e63c4bed236e63af5, and it seems wrong that `Assert.Throws<TActual>()` should return `TActual?` and not `TActual`. Shouldn't `Assert.Throws()` always throw a test assertion exception if the test delegate does not throw the expected `TActual` exception? I don't think it's right that I should need to explicitly test that `ex` is not `null` after a call to `Assert.Throws()`.

**Comment by manfred-brands** [comment]
@egraff For an explanation see: #3717
In our company we dropped the use of ReSharper completely in favour of the Roslyn Analyzers. ReSharper slowed VS down too much and didn't add enough for us to live with that.

**Comment by egraff** [comment]
Got it, thanks! 🙂

**Comment by jnm2** [comment]
@egraff The intellisense docs explain when it can return null rather than throwing an exception. See https://github.com/nunit/nunit/issues/3717.

**Comment by Rabadash8820** [comment]
@jnm2 Any updates on this? I just ran into this situation after updating my Unity game project to Unity 2021.2, which now supports C#8. I'm starting to take advantage of nullable reference types both in my code and my unit tests, and it would be nice not to have a bunch of extra `!` characters in my unit tests right after a call to `Assert.IsNotNull`. 😛

**Comment by smkanadl** [comment]
@Rabadash8820 Do you have the `NUnit.Analyzers` NuGet package installed in your test project? It does a very decent job in suppressing most of the typical nullable warnings.

**Comment by Rabadash8820** [comment]
I do not, first I've heard of it. 🤔 Let me give that a try...

**Comment by jnm2** [comment]
@Rabadash8820 The nullability annotation in the `Assert.IsNotNull` API is what we think is the most accurate (given the constraints of C#'s expressivity) and it is considered final. The reason for it being the way it is is given in the XML documentation that shows up in IntelliSense for this method.

**Comment by jnm2** [comment]
This comes up so routinely that I think the NUnit package should begin depending on the NUnit.Analyzers package to pull it in everywhere NUnit is used. (There are various methods of opting out that would then be available which we could document.)

**Comment by OsirisTerje** [comment]
Been thinking that the adapter itself should depend on both of these.  The adapter without nunit is sort of.... well completely useless.  The analyzer is now so good it should be included, agree nunit should depend on it.

**Comment by Rabadash8820** [comment]
@jnm2 Maybe I'm misunderstanding you... I just updated to the latest version of NUnit (3.13.2). If I remove the NUnit.Analyzers package, then the code below shows a null warning. If I add back the analyzers package, then the warning goes away. In both cases, however, I do not see any `[NotNull*]` attributes on the `anObject` parameter of `IsNotNull`, nor XML comments explaining why.

```cs
bool isNull = true;
object? derp = isNull ? null : new();
Assert.IsNotNull(derp);
derp.ToString();    // CS8602 Dereference of a possibly null reference.
```

**Comment by jnm2** [comment]
@Rabadash8820 Does this not show up in your IDE?

![image](https://user-images.githubusercontent.com/8040367/144155144-cd3835f0-0cf5-46f5-9853-f74d5a7b4d90.png)

![Uploading image.png…]()

**Comment by Rabadash8820** [comment]
It does. Is that the message to which you were referring? Not throwing an exception is different from convincing the compiler that a nullable ref is now _certainly_ non-null.

**Comment by jnm2** [comment]
It's not non-null after the call unless Assert.IsNotNull throws an exception if the variable is null, which it does not always.

```cs
Assert.Multiple(() =>
{
    object? x = null;
    Assert.IsNotNull(x);
    x.ToString(); // Blows up at runtime here with NRE.
});
```

**Comment by Rabadash8820** [comment]
Hm, ok I think I understand now. So basically, none of the following attributes will help:
- `NotNull`: because `IsNotNull` returning (not throwing) does _not_ guarantee that the value is non-null
- `NotNullWhen`: because `IsNotNull` returns `void`, so there's nothing to pass to that attribute's constructor
- `NotNullIfNotNull`: because, again, `IsNotNull` returns `void`, and anyway we would want to validate null state of the _provided_ value, not the return value

@jnm2 Sounds like you've already considered all this. 😛 And great work again on the analyzer package; that definitely meets my immediate needs. Any updates on what this will look like in NUnit v4, or when that will be coming out? I didn't see anything nullable refs in #3325.

**Comment by jnm2** [comment]
@Rabadash8820 Right, there's no way to annotate a method as having behavior dependent upon whether the call site is inside a lambda argument to some other method.

We haven't come across anything to discuss specific to NUnit v4 having to do with nullability annotations other than having the NUnit.Analyzers package be a dependency of the NUnit package. The remaining annotations (IConstraint mainly IIRC) will be done in NUnit v3 unless we stop generally adding new things to v3. I could foresee v4 becoming related to nullability annotations if we want to add annotations that introduce a lot of build warnings and we don't want to affect v3 with that.

**Comment by OsirisTerje** [comment]
@jnm2 What is the current state of this rather long-living issue?   I haven't read everything, but do you think we can move forward on this?  
It would be awesome if you could write up a small summary here.  
Also, can NUnit as a library handle the nullability property?  
If not, do we need to compile one with and one without?  
If all are added, what would be the effect when using older C# versions?

**Comment by jnm2** [comment]
Off the top of my head, I think almost all of NUnit has been annotated except the Constraints namespace which had some thorny issues. We can track the state by seeing how many files have `#nullable enable` in them. Eventually, we could flip this around with `<Nullable>enable</Nullable>` in the project file and `#nullable disable` where necessary.

Nullability properties are already shipping in the library. They are backwards-compatible, producing no requirements when consumed from older versions of C#.

**Comment by OsirisTerje** [comment]
It's due time to set nullable enabled in the project files, and I agree, we can set the remaining to disable in their files.

**Comment by OsirisTerje** [comment]
I've made another attempt at just resolving stuff in the framework.test project, with a few fixes in the framework itself.    That didn't go very well......

We need to 
1:  Take return values first.    
1.1: For collections, where they have nothing in return, we should not return them as null, but instead return empty collections.  That should be the preferred method for all.
1.2: To ease integration where these are used, we should replace named types with the `var` keyword, then acceptance of possible nulls is easier.
2.  We need to fix one file at a time.  Doing it per project just gets overwhelming. So we need to add `#nullable enable` at the top of the cs files, as @jnm2 mentions above.  It will take time, but it looks like the best way of approaching this.

**Comment by manfred-brands** [comment]
@OsirisTerje I have been doing a lot of nullable conversion for work. But most of these were from ReSharper attributes (CanBeNull/NotNull) to C#8 nullable where some regular expressions did wonders.
One thing to be aware off is that if compiling for older frameworks (anything before NET5), the referenced assemblies are not annotated. This will result in the compiler not detecting that if `string.IsNullOrEmpty(s)` returns `false` then `s` is not null.
We have made annotated reference assemblies for older frameworks, using the [ReferenceAssmblyAnnotator](https://www.nuget.org/packages/TunnelVisionLabs.ReferenceAssemblyAnnotator/1.0.0-alpha.160) nuget package.

I don't like using `var` unless the type is clearly specified on the right hand side. Doing reviews outside VS are much easier that way as one can determine the type from looking at the code without having to look up the method/property called.

There are 431 .cs files in the framework, but 215 already have `#nullable enable` in them, so we are already half-way!
I can take this one on if you like.

**Comment by manfred-brands** [comment]
I assigned this to myself.

**Comment by manfred-brands** [comment]
I had a good session today, converting almost all files to nullable.

It highlights some design issues:

1. A base class declares a property (or worse protected field) that is expected to be set by a derived class.
If that property is declared non-null, this result in errors as the base class doesn't set it. If declared nullable, all usages result in errors as they assume it to be non-null. There is no way to guarantee that a derived class actually sets that field/property.

For `Constraint` I solved this by declaring the property as `abstract` in the base class and making appropriate changes, e.g. in `NullConstraint`. This showed where derived classes didn't do what they were supposed to do. Luckily only in TestFixtures.

2. A class that set fields in one public method (e.g. `WorkItem.InitializeContext`) that is expected to be set in others.
Again, this order cannot be enforced by the compiler. 
We could add `Guard.OperationValid(Context != null, "InitializeContect not called")` but the `Context` property is used in derived classes as well. 

3. The `Context` class itself has lots of fields that are assumed to be set and not null. Most are only set in the `WorkItem.RunOnCurrentThread` method, but are assumed to be set everywhere. Same problem as 2.

4. `TypeConstraint` has a field `actualType` which is only ever used inside the `Apply` method as a set and report. It could be removed, but that would break any code, outside NUnit, relying on the field.

I need to go over a few more things.

**Comment by manfred-brands** [comment]
Converting the nunit.framework.tests to nullable is likely to result in more changes.
